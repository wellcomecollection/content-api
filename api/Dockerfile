# Build this with the repo root as the build context
FROM public.ecr.aws/docker/library/node:18

WORKDIR /app

COPY ./package.json ./yarn.lock ./
COPY ./api ./api

WORKDIR api
RUN yarn install --frozen-lockfile && yarn cache clean

CMD ["yarn", "start"]
