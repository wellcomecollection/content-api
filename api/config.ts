import { URL } from 'url';
import { z } from 'zod';

const environmentSchema = z.object({
  PUBLIC_ROOT_URL: z
    .string()
    .url()
    .default('https://api.wellcomecollection.org/content/v0'),
});
const environment = environmentSchema.parse(process.env);

// This configuration is exposed via the public healthcheck endpoint,
// so be careful not to expose any secrets here.
const config = {
  pipelineDate: '2025-07-30',
  addressablesIndex: 'addressables',
  articlesIndex: 'articles',
  eventsIndex: 'events',
  venuesIndex: 'venues',
  publicRootUrl: new URL(environment.PUBLIC_ROOT_URL),
};

export type Config = typeof config;

export const getConfig = () => config;
