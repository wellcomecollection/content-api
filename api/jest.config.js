module.exports = {
  displayName: "API",
  preset: "ts-jest",
  testEnvironment: "node",
  setupFilesAfterEnv: ["<rootDir>/test/setup.js"],
};
