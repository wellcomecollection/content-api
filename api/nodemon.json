{
  "watch": ["src"],
  "ext": "ts",
  "exec": "tsx ./server.ts"
}
