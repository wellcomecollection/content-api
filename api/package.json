{
  "name": "@weco/content-api",
  "version": "0.1.0",
  "description": "An API for serving content, such as articles",
  "license": "MIT",
  "scripts": {
    "start": "NODE_ENV=production ts-node-transpile-only ./server.ts",
    "dev": "AWS_PROFILE=catalogue-developer NODE_ENV=development nodemon ./server.ts",
    "test": "jest"
  },
  "dependencies": {
    "@elastic/elasticsearch": "^8.6.0",
    "@types/express": "^4.17.21",
    "@weco/content-common": "*",
    "express": "^4.21.0",
    "express-async-handler": "^1.2.0",
    "luxon": "^3.4.4",
    "morgan": "^1.10.0",
    "ts-node": "^10.9.1",
    "typescript": "^5.6.2"
  },
  "devDependencies": {
    "@tsconfig/node20": "20.1.4",
    "@types/jest": "^29.5.13",
    "@types/luxon": "^3.4.2",
    "@types/morgan": "^1.9.3",
    "@types/node": "^20.16.5",
    "@types/supertest": "^2.0.12",
    "jest": "^29.7.0",
    "jest-extended": "^4.0.2",
    "nodemon": "^2.0.20",
    "supertest": "^6.3.3",
    "ts-jest": "^29.2.5"
  }
}
