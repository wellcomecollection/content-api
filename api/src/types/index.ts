import * as prismic from "@prismicio/client";

import { TransformedArticle, TransformedContributor } from "./articles";
export type { TransformedArticle, TransformedContributor };

export type Clients = {
  prismic: prismic.Client;
};

export type ResultList<Result> = {
  type: "ResultList";
  results: Result[];
};

export type ContentType = {
  id: string;
};
