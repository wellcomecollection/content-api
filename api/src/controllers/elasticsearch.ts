import { Client } from "@elastic/elasticsearch";
import * as prismic from "@prismicio/client";
import {
  Article,
  ArticlePrismicDocument,
  Clients,
  ContentType,
} from "../types";
import { CountResponse } from "@elastic/elasticsearch/lib/api/types";
import { getConfig } from "../../config";
import { graphQueryArticles } from "../helpers/articles";
import { transformArticle } from "../transformers/article";
import { HttpError } from "./error";
import { PaginationQueryParameters } from "./pagination";
import { RequestHandler } from "express";
import { OnDropDocument } from "@elastic/elasticsearch/lib/helpers";

const indexName = getConfig().contentsIndex;

type QueryParams = {
  query?: string;
} & PaginationQueryParameters;

type Handler = RequestHandler<
  never,
  { count: CountResponse },
  never,
  QueryParams
>;

// TODO remove?
const addIndex = async (elasticClient: Client, indexName: string) => {
  const exists = await elasticClient.indices.exists({ index: indexName });

  if (!exists) {
    await elasticClient.indices.create({ index: indexName });
    console.log(indexName, "was created");
  } else {
    console.log("Index", indexName, "already exists");
  }
};

const bulkIndexDocuments = async (elasticClient: Client, docs: Article[]) => {
  const operations = docs.flatMap((doc) => {
    return [
      {
        display: doc,
        query: { id: doc.id },
      },
    ];
  });

  let failures: OnDropDocument<{ display: any; query: any }>[] = [];
  await elasticClient.helpers.bulk({
    datasource: operations,
    onDocument(doc) {
      return {
        index: { _index: indexName, _id: doc.display.id },
      };
    },
    onDrop: (fail) => {
      failures.push(fail);
    },
  });

  if (failures.length !== 0) {
    const failedServices = new Set(
      failures.map(({ document }) => document.display)
    );
    console.error(
      `Failed to ingest documents from ${failedServices.size} services: ${[
        ...failedServices,
      ].join(", ")}`
    );
    throw new Error(JSON.stringify(failures));
  }

  const count = await elasticClient.count({ index: indexName });
  return count;
};

// Review what prismic function we use here.
// getAllByType (no pagination blocker but only one type at a time)
// or getByType (which allows more than one type at a time).
// Review if predicates are still required
const getPrismicDocuments = async <T>({
  prismicClient,
  contentTypes,
}: {
  prismicClient: prismic.Client;
  contentTypes: ContentType[];
}): Promise<T> => {
  const getDocuments = await prismicClient.getAllByType(contentTypes[0], {
    graphQuery: graphQueryArticles,
    predicates: [prismic.predicate.any("document.type", contentTypes)],
  });

  return getDocuments as T;
};

const elasticsearchController = (clients: Clients): Handler => {
  return async (req, res) => {
    try {
      // Fetch all articles and webcomics from Prismic
      const searchResponse = await getPrismicDocuments<
        ArticlePrismicDocument[]
      >({
        prismicClient: clients.prismic,
        contentTypes: ["articles", "webcomics"],
      });

      if (searchResponse) {
        // Transform all articles and webcomics
        // As this is for testing, we don't need all 900 articles
        const transformedResponse = searchResponse
          .slice(0, 40)
          .map((result) => transformArticle(result));

        // Bulk send them to Elasticsearch
        const bulkHelper = await bulkIndexDocuments(
          clients.elastic,
          transformedResponse
        );

        res.status(200).json({ count: bulkHelper });
      } else {
        throw new HttpError({
          status: 404,
          label: "No results found",
        });
      }
    } catch (error) {
      // TODO handle this better, what would we want here?
      console.log({ error });
      throw error;
    }
  };
};

export default elasticsearchController;
