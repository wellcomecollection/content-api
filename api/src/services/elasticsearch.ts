import { Client } from "@elastic/elasticsearch";

// const isProduction = process.env.NODE_ENV === "production";

export const getElasticClient = async (): Promise<Client> => {
  const client = await new Client({
    node: "https://content.es.eu-west-1.aws.found.io:9243",
    auth: {
      username: "dev",
      password: process.env.ELASTICSEARCH_PASSWORD || "",
    },
  });
  return client;
};
