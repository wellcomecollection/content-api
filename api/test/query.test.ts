import { Client as ElasticClient } from "@elastic/elasticsearch";
import { SearchRequest } from "@elastic/elasticsearch/lib/api/types";
import supertest from "supertest";
import createApp from "../src/app";
import { URL, URLSearchParams } from "url";

describe("articles query", () => {
  // The purpose of this test is as a smoke test for the question,
  // "do we understand how we map a given query into an ES request?"
  it("makes the expected query to ES for a given set of query parameters", async () => {
    const aggregations = "format,contributors.contributor";
    const format = "test-format";
    const contributor = "test-contributor";
    const pageSize = 42;
    const page = 9;
    const dateFrom = "2022-02-22";
    const dateTo = "2023-03-23";
    const sort = "publicationDate";
    const sortOrder = "asc";
    const query = "henry wellcome";

    const params = new URLSearchParams({
      aggregations,
      format,
      page,
      pageSize,
      sort,
      sortOrder,
      query,
      "publicationDate.from": dateFrom,
      "publicationDate.to": dateTo,
      "contributors.contributor": contributor,
    } as unknown as Record<string, string>);
    const esRequest = await elasticsearchRequestForURL(
      `/articles?${params.toString()}`
    );

    expect(esRequest.from).toBe((page - 1) * pageSize);
    expect(esRequest.size).toBe(pageSize);
    expect(esRequest.aggregations).toContainAllKeys(aggregations.split(","));

    expect(JSON.stringify(esRequest.query?.bool?.must)).toInclude(query);
    expect(JSON.stringify(esRequest.query?.bool?.filter)).toInclude(dateFrom);
    expect(JSON.stringify(esRequest.query?.bool?.filter)).toInclude(dateTo);
    expect(JSON.stringify(esRequest.post_filter?.bool?.filter)).toInclude(
      contributor
    );
    expect(JSON.stringify(esRequest.post_filter?.bool?.filter)).toInclude(
      format
    );

    expect(esRequest).toMatchSnapshot();
  });
});

const elasticsearchRequestForURL = async (
  url: string
): Promise<SearchRequest> => {
  const searchSpy = jest.fn();
  const app = createApp(
    { elastic: { search: searchSpy } as unknown as ElasticClient },
    {
      pipelineDate: "2222-22-22",
      articlesIndex: "test",
      eventsIndex: "",
      publicRootUrl: new URL("http://test.test/test"),
    }
  );
  const api = supertest.agent(app);
  await api.get(url);
  return searchSpy.mock.lastCall[0] as SearchRequest;
};
