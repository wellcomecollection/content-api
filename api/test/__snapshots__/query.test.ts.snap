// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`articles query makes the expected query to ES for a given set of query parameters 1`] = `
{
  "_source": [
    "display",
  ],
  "aggregations": {
    "format": {
      "terms": {
        "field": "aggregatableValues.format",
        "size": 20,
      },
    },
  },
  "from": 336,
  "index": "test",
  "post_filter": {
    "bool": {
      "filter": [
        {
          "terms": {
            "filter.formatId": [
              "test-format",
            ],
          },
        },
      ],
    },
  },
  "query": {
    "bool": {
      "filter": [
        {
          "terms": {
            "filter.contributorIds": [
              "test-contributor",
            ],
          },
        },
        {
          "range": {
            "filter.publicationDate": {
              "gte": "2022-02-22T00:00:00.000Z",
              "lte": "2023-03-23T00:00:00.000Z",
            },
          },
        },
      ],
      "must": {
        "multi_match": {
          "fields": [
            "id",
            "query.title.*^100",
            "query.contributors.*^10",
            "query.contributors.keyword^100",
            "query.standfirst.*^10",
            "query.body.*",
            "query.caption.*",
            "query.series.id",
            "query.series.title.*^80",
            "query.series.contributors*^8",
            "query.series.contributors.keyword^80",
          ],
          "minimum_should_match": "-25%",
          "operator": "or",
          "query": "henry wellcome",
          "type": "cross_fields",
        },
      },
    },
  },
  "size": 42,
  "sort": [
    {
      "query.publicationDate": {
        "order": "asc",
      },
    },
    {
      "query.publicationDate": {
        "order": "desc",
      },
    },
  ],
}
`;
